### PR TITLE
Added API & Vanish

### PR DESCRIPTION
**Pull request by alexdev03**
I have implemented the VelocitabAPI, which includes methods for changing a player's name, retrieving TabPlayer information, and handling vanish functionality. The VanishIntegration interface will need to be implemented by a vanish plugin and registered with the VelocitabAPI. The logic for the vanish feature has been implemented but has not yet been tested. Once this pull request is merged, I will connect a vanish plugin to the VelocitabAPI to test its functionality. Additionally, I have made improvements to the tabplayer cache. I have also added a new command, /velocitab name, which allows for testing custom player names on the tablist. In the future, I plan to add support for custom nametags as well.

### CODE REVIEW

File: src/main/java/net/william278/velocitab/vanish/DefaultVanishIntegration.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package net.william278.velocitab.vanish;
+
+public class DefaultVanishIntegration implements VanishIntegration {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/DefaultVanishIntegration.java:22]: lets make this a `final` class so people aren't weird and try and extend this

File: src/main/java/net/william278/velocitab/vanish/VanishIntegration.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package net.william278.velocitab.vanish;
+
+public interface VanishIntegration {
+
+    boolean canSee(String name, String otherName);
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/VanishIntegration.java:24]:
```suggestion
    boolean canSee(@NotNull String name, @NotNull String otherName);
```

File: src/main/java/net/william278/velocitab/vanish/DefaultVanishIntegration.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    @Override
+    public boolean canSee(String name, String otherName) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/DefaultVanishIntegration.java:24]:
```suggestion
    public boolean canSee(@NotNull String name, @NotNull String otherName) {
```

File: src/main/java/net/william278/velocitab/vanish/DefaultVanishIntegration.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        return true;
+    }
+
+    @Override
+    public boolean isVanished(String name) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/DefaultVanishIntegration.java:29]:
```suggestion
    public boolean isVanished(@NotNull String name) {
```

File: src/main/java/net/william278/velocitab/vanish/VanishIntegration.java [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+    boolean isVanished(String name);
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/VanishIntegration.java:26]:
```suggestion
    boolean isVanished(@NotNull String name) {
```

File: src/main/java/net/william278/velocitab/vanish/VanishManager.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+import java.util.Objects;
+import java.util.Optional;
+
+public class VanishManager {
+
+    private final Velocitab plugin;
+    private VanishIntegration integration;
+
+    public VanishManager(Velocitab plugin) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/VanishManager.java:35]:
```suggestion
    public VanishManager(@NotNull Velocitab plugin) {
```

File: src/main/java/net/william278/velocitab/vanish/DefaultVanishIntegration.java (hunk 1 above)
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/DefaultVanishIntegration.java:22]: Let's add a protected zero-args constructor here so people can't instantiate this outside this package either

File: src/main/java/net/william278/velocitab/vanish/VanishManager.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+    public void setIntegration(@NotNull VanishIntegration integration) {
+        this.integration = integration;
+    }
+
+    public @NotNull VanishIntegration getIntegration() {
+        return integration;
+    }
+
+    public boolean canSee(String name, String otherName) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/VanishManager.java:48]:
```suggestion
    public boolean canSee(@NotNull String name, @NotNull String otherName) {
```

File: src/main/java/net/william278/velocitab/vanish/VanishManager.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        return integration.canSee(name, otherName);
+    }
+
+    public boolean isVanished(String name) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/VanishManager.java:52]:
```suggestion
    public boolean isVanished(@NotNull String name) {
```

File: src/main/java/net/william278/velocitab/vanish/VanishManager.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        return integration.isVanished(name);
+    }
+
+    public void vanishPlayer(Player player) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/VanishManager.java:56]:
```suggestion
    public void vanishPlayer(@NotNull Player player) {
```

File: src/main/java/net/william278/velocitab/vanish/VanishManager.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        Optional<TabPlayer> tabPlayer = plugin.getTabList().getTabPlayer(player);
+
+        if(tabPlayer.isEmpty()) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/VanishManager.java:59]:
```suggestion
        if (tabPlayer.isEmpty()) {
```

File: src/main/java/net/william278/velocitab/vanish/VanishManager.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        return integration.canSee(name, otherName);
+    }
+
+    public boolean isVanished(String name) {
+        return integration.isVanished(name);
+    }
+
+    public void vanishPlayer(Player player) {
+        Optional<TabPlayer> tabPlayer = plugin.getTabList().getTabPlayer(player);
+
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/VanishManager.java:58]: `final Optional<TabPlayer>`
plus, remove the whitespace below

File: src/main/java/net/william278/velocitab/vanish/VanishManager.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+    public VanishManager(Velocitab plugin) {
+        this.plugin = plugin;
+        setIntegration(new DefaultVanishIntegration());
+    }
+
+    public void setIntegration(@NotNull VanishIntegration integration) {
+        this.integration = integration;
+    }
+
+    public @NotNull VanishIntegration getIntegration() {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/VanishManager.java:44]:
```suggestion
    @NotNull
    public VanishIntegration getIntegration() {
```

File: src/main/java/net/william278/velocitab/vanish/VanishManager.java [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+        if(tabPlayer.isEmpty()) {
+            return;
+        }
+
+        plugin.getTabList().vanishPlayer(tabPlayer.get());
+        plugin.getScoreboardManager().ifPresent(scoreboardManager -> scoreboardManager.vanishPlayer(player));
+    }
+
+    public void unvanishPlayer(Player player) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/VanishManager.java:67]:
```suggestion
    public void unvanishPlayer(@NotNull Player player) {
```

File: src/main/java/net/william278/velocitab/vanish/VanishManager.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+        Optional<TabPlayer> tabPlayer = plugin.getTabList().getTabPlayer(player);
+
+        if(tabPlayer.isEmpty()) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/VanishManager.java:70]:
```suggestion
        if (tabPlayer.isEmpty()) {
```

File: src/main/java/net/william278/velocitab/vanish/VanishManager.java [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+            return;
+        }
+
+        plugin.getTabList().vanishPlayer(tabPlayer.get());
+        plugin.getScoreboardManager().ifPresent(scoreboardManager -> scoreboardManager.vanishPlayer(player));
+    }
+
+    public void unvanishPlayer(Player player) {
+        Optional<TabPlayer> tabPlayer = plugin.getTabList().getTabPlayer(player);
+
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/VanishManager.java:69]: `final Optional<TabPlayer>`
plus, remove the whitespace below

File: src/main/java/net/william278/velocitab/packet/ScoreboardManager.java [hunk 16]
```diff
@@ -74,14 +74,72 @@ public void resetCache(@NotNull Player player) {
         }
     }
 
+    public void vanishPlayer(Player player) {
+        if (!plugin.getSettings().doNametags()) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/packet/ScoreboardManager.java:78]: Is this related to the nametags feature? Can prob remove this check

File: src/main/java/net/william278/velocitab/packet/ScoreboardManager.java [hunk 17]
```diff
@@ -74,14 +74,72 @@ public void resetCache(@NotNull Player player) {
...
+            if (!canSee) {
+                return;
+            }
+
+            dispatchPacket(packet, connected);
+        }));
+    }
+
+    public void unvanishPlayer(Player player) {
+        if (!plugin.getSettings().doNametags()) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/packet/ScoreboardManager.java:104]: Is this related to the nametags feature? Can prob remove this check

**Review by WiIIiam278 — CHANGES_REQUESTED** [review]
Few nits but looking good

File: src/main/java/net/william278/velocitab/api/VelocitabAPI.java [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+    @ApiStatus.Internal
+    protected VelocitabAPI(@NotNull Velocitab plugin) {
+        this.plugin = plugin;
+    }
+
+    /**
+     * Entrypoint to the HuskSync API - returns an instance of the API
+     *
+     * @return instance of the HuskSync API
+     * @since 1.5.2
```
[the remark this answers is not in the document]
Comment (alexdev03) [reply src/main/java/net/william278/velocitab/api/VelocitabAPI.java:55]: nostalgic

File: src/main/java/net/william278/velocitab/packet/ScoreboardManager.java (hunk 16 above)
Comment (alexdev03) [reply src/main/java/net/william278/velocitab/packet/ScoreboardManager.java:78]: Yes maybe sorting

File: src/main/java/net/william278/velocitab/api/VelocitabAPI.java [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+
+import java.util.Optional;
+
+@SuppressWarnings("unused")
+
+/**
+ * The Velocitab API class.
+ * </p>
+ * Retrieve an instance of the API class via {@link #getInstance()}.
+ */ public class VelocitabAPI {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/api/VelocitabAPI.java:40]: formatting mistake on this javadoc comment here

File: src/main/java/net/william278/velocitab/tab/PlayerTabList.java [hunk 20]
```diff
@@ -342,4 +384,33 @@ public void proxyReload(@NotNull ProxyReloadEvent event) {
...
+                return;
+            }
+
+            if (!plugin.getVanishManager().canSee(p.getPlayer().getUsername(), tabPlayer.getPlayer().getUsername())) {
+                p.getPlayer().getTabList().removeEntry(tabPlayer.getPlayer().getUniqueId());
+            }
+        });
+    }
+
+    public void unvanishPlayer(@NotNull TabPlayer tabPlayer) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/tab/PlayerTabList.java:400]: let's change this to `unVanishPlayer`

File: src/main/java/net/william278/velocitab/api/VelocitabAPI.java [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+     * Vanishes the player by hiding them from the tab list and scoreboard if enabled.
+     *
+     * @param player The player to vanish
+     */
+    public void vanishPlayer(@NotNull Player player) {
+        plugin.getVanishManager().vanishPlayer(player);
+    }
+
+    /**
+     * Unvanishes the given player by showing them in the tab list and scoreboard if enabled.
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/api/VelocitabAPI.java:159]: jd comment: "Un-vanishes" instead of "Unvanishes"

File: src/main/java/net/william278/velocitab/commands/VelocitabCommand.java [hunk 22]
```diff
@@ -20,15 +20,23 @@
...
 import com.mojang.brigadier.builder.LiteralArgumentBuilder;
+import com.mojang.brigadier.builder.RequiredArgumentBuilder;
 import com.velocitypowered.api.command.BrigadierCommand;
 import com.velocitypowered.api.command.CommandSource;
+import com.velocitypowered.api.proxy.Player;
 import net.kyori.adventure.text.Component;
 import net.kyori.adventure.text.format.TextColor;
 import net.william278.desertwell.about.AboutMenu;
 import net.william278.velocitab.Velocitab;
+import net.william278.velocitab.api.VelocitabAPI;
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/commands/VelocitabCommand.java:34]: Can we cleanup some of these now-unused imports? CTRL+ALT+O in intelliJ after selecting the root package

File: src/main/java/net/william278/velocitab/config/Placeholder.java
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/config/Placeholder.java:1]: Further down in this class we can remove the `formatSortableInt` method now we've got the SortingManager

File: src/main/java/net/william278/velocitab/packet/ScoreboardManager.java [hunk 23]
```diff
@@ -74,14 +74,71 @@ public void resetCache(@NotNull Player player) {
         }
     }
 
+    public void vanishPlayer(Player player) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/packet/ScoreboardManager.java:77]:
```suggestion
    public void vanishPlayer(@NotNull Player player) {
```

File: src/main/java/net/william278/velocitab/packet/ScoreboardManager.java [hunk 24]
```diff
@@ -74,14 +74,71 @@ public void resetCache(@NotNull Player player) {
...
+
+            if (!canSee) {
+                return;
+            }
+
+            dispatchPacket(packet, connected);
+        }));
+    }
+
+    public void unvanishPlayer(Player player) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/packet/ScoreboardManager.java:103]:
```
    public void unVanishPlayer(@NotNull Player player) {
```

File: src/main/java/net/william278/velocitab/packet/ScoreboardManager.java [hunk 25]
```diff
@@ -122,16 +179,31 @@ public void resendAllNameTags(Player player) {
                 .map(RegisteredServer::getPlayersConnected)
                 .flatMap(Collection::stream)
                 .toList();
+
+        List<String> roles = new ArrayList<>();
+
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/packet/ScoreboardManager.java:184]: `final List<String> roles` and remove the newline after it

File: src/main/java/net/william278/velocitab/packet/ScoreboardManager.java [hunk 26]
```diff
@@ -107,7 +164,7 @@ public void updateRole(@NotNull Player player, @NotNull String role) {
 
 
     public void resendAllNameTags(Player player) {
-        if (!plugin.getSettings().doNametags()) {
+        if (!plugin.getSettings().isSortPlayers()) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/packet/ScoreboardManager.java:167]: No I think we want to keep `doNametags()` for this one!

File: src/main/java/net/william278/velocitab/tab/PlayerTabList.java [hunk 27]
```diff
@@ -204,19 +240,25 @@ public void updatePlayer(@NotNull TabPlayer tabPlayer) {
...
             }
 
-            players.forEach(player ->
-                    player.getPlayer().getTabList().getEntries().stream()
-                            .filter(e -> e.getProfile().getId().equals(tabPlayer.getPlayer().getUniqueId())).findFirst()
-                            .ifPresent(entry -> entry.setDisplayName(displayName)));
-        });
+            boolean isVanished = plugin.getVanishManager().isVanished(tabPlayer.getPlayer().getUsername());
 
+            players.forEach(player -> {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/tab/PlayerTabList.java:252]: `final boolean isVanished` and remove newline

File: src/main/java/net/william278/velocitab/vanish/VanishManager.java [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+        final Optional<TabPlayer> tabPlayer = plugin.getTabList().getTabPlayer(player);
+        if (tabPlayer.isEmpty()) {
+            return;
+        }
+
+        plugin.getTabList().vanishPlayer(tabPlayer.get());
+        plugin.getScoreboardManager().ifPresent(scoreboardManager -> scoreboardManager.vanishPlayer(player));
+    }
+
+    public void unvanishPlayer(@NotNull Player player) {
```
Comment (WiIIiam278) [inline src/main/java/net/william278/velocitab/vanish/VanishManager.java:67]: `unVanishPlayer`

File: src/main/java/net/william278/velocitab/packet/ScoreboardManager.java (hunk 26 above)
Comment (alexdev03) [reply src/main/java/net/william278/velocitab/packet/ScoreboardManager.java:167]: The sorting system is handled with teams. Maybe the problem is with the method name, I would change that to resendAllTeams

Comment (WiIIiam278) [reply src/main/java/net/william278/velocitab/packet/ScoreboardManager.java:167]: You should check the Settings#doScoreboardTeamsPackets setting instead

**Review by WiIIiam278 — APPROVED** [review]
